ENT-10284 Performance optimise deserialisation

rick-r3 opened the pull request:
Finally tick off a long-standing TODO.

DeserializationInput will only parse the schema elements of the payload if the we have never seen one or more types before, so over time we stop parsing the schema which can be expensive.  Some results from microbenchmarks with a SignedTransaction blob we have used in other tests.

Benchmark test deserialises a SignedTransaction blob.  Improvements as follows:

4x vs. current 4.11
9x vs. 4.10 (and earlier)

4.9 			29442 nanoseconds
4.10 		29442 nanoseconds
4.11 		14680 nanoseconds
4.11 + PR   	  3318 nanoseconds

Expanding that to deserialise the WiredTransaction, inputs, outputs and commands:

2.5x vs. current 4.11
5x vs. 4.10 (and earlier)

4.9 			178016 nanoseconds
4.10 		178016 nanoseconds
4.11 		87566 nanoseconds
4.11 + PR   	 35269 nanoseconds

rick-r3 committed:
Lazy extraction of schema from binary.  Not lazy passing around yet.

rick-r3 committed:
Still returning decoder to pool too early, so currently not re-using them.

rick-r3 committed:
Fixed pooling and laziness interactions

rick-r3 committed:
Works with DJVM

rick-r3 committed:
Performance test

chrisr3 commented on core-deterministic/src/main/kotlin/net/corda/core/internal/LazyPool.kt:
```diff
@@ -0,0 +1,41 @@
+ * @param newInstance The function to call to lazily newInstance a pooled resource.
+ */
+@Suppress("unused")
+@KeepForDJVM
```
Does this actually _need_ annotating as `@KeepForDJVM`? IIRC that annotation is to prevent ProGuard from throwing a class away as unused.

rick-r3 replied:
Without it I get class missing issues and couldn't figure out any other way to solve it.   Took me ages to come to the conclusion to try this.

chrisr3 replied:
OK, I guess it does then :grin:. The `deterministic` artifacts are created in 4 steps:
- An initial ProGuard step to keep everything marked as `@KeepForDJVM`, and discard anything which this subset doesn't need.
- A `jar-filter` step to remove classes, methods etc that have been annotated as `@DeleteForDJVM` or `@StubForDJVM`.
- A repeat of the ProGuard step on the `@KeepForDJVM` classes to tidy everything up.
- Finally, rewrite all of the Kotlin classes' metadata to describe their new contents correctly.

You can gain a better insight into what Gradle is doing here by enabling its "info" or "debug" logging, i.e. `-i` or `-d` command line parameters.

rick-r3 replied:
Of course!  My brain was failing to work that out.

chrisr3 commented on serialization/src/main/kotlin/net/corda/serialization/internal/amqp/SerializerFactory.kt:
```diff
@@ -5,7 +5,17 @@ import java.io.NotSerializableException
+class SerializationSchemas(val resolveSchema: () -> Pair<Schema, TransformsSchema>) {
+        constructor(schema: Schema, transforms: TransformsSchema) : this({ schema to transforms })
+
+        private val resolvedSchema: Pair<Schema, TransformsSchema> by lazy { resolveSchema() }
```
Maybe

```kotlin
by lazy(::resolveSchema)
```

rick-r3 replied:
Hmmm.... Doesn't like it.

<img width="887" alt="image" src="https://github.com/corda/corda/assets/23558791/ca24500d-55b8-4532-9b6b-221f449453fd">

chrisr3 replied:
_Doh!_ Try:
```kotlin
by lazy(resolveSchema)
```
which also means that `resolveSchema` no longer needs to be a property:
```kotlin
class SerializationSchemas(resolveSchema: () -> Pair<Schema, TransformsSchema>) {
```

rick-r3 replied:
Of course!  My brain was failing to work that out.

chrisr3 commented on serialization/src/main/kotlin/net/corda/serialization/internal/amqp/Envelope.kt:
```diff
@@ -13,7 +18,13 @@ import org.apache.qpid.proton.codec.DescribedTypeConstructor
-data class Envelope(val obj: Any?, val schema: Schema, val transformsSchema: TransformsSchema) : DescribedType {
+class Envelope(val obj: Any?, private val resolveSchema: () -> Pair<Schema, TransformsSchema>) : DescribedType {
+
+    val resolvedSchema: Pair<Schema, TransformsSchema> by lazy { resolveSchema() }
```
Also
```kotlin
by lazy(::resolveSchema)
```

rick-r3 replied:
Same as above.  Kotlin doesn't like it.

chrisr3 replied:
_Meh!_ Try `by lazy(resolveSchema)` again. Which also no longer needs to be a property:
```kotlin
class Envelope(val obj: Any?, resolveSchema: () -> Pair<Schema, TransformsSchema>) : DescribedType {
```
IntelliJ also recommends this for line#76:
```kotlin
return Envelope(list[BLOB_IDX]) { list[SCHEMA_IDX] as Schema to transformSchema }
```

rick-r3 replied:
Of course!  My brain was failing to work that out.

chrisr3 commented on serialization/src/main/kotlin/net/corda/serialization/internal/amqp/DeserializationInput.kt:
```diff
@@ -155,10 +179,10 @@ class DeserializationInput constructor(
-            type = clazz,
-            context = context
+                obj = redescribe(envelope.obj, clazz),
+                schema = SerializationSchemas({ envelope.resolvedSchema }),
```
```kotlin
SerializationSchemas(envelope::resolvedSchema),
```

chrisr3 commented on serialization/src/main/kotlin/net/corda/serialization/internal/amqp/DeserializationInput.kt:
```diff
@@ -72,17 +74,32 @@ class DeserializationInput constructor(
 
+        private val decoderPool = LazyPool<DecoderImpl> {
+            val decoder = DecoderImpl().apply {
+                this.register(Envelope.DESCRIPTOR, Envelope.FastPathConstructor(this))
+                this.register(Schema.DESCRIPTOR, Schema)
+                this.register(Descriptor.DESCRIPTOR, Descriptor)
+                this.register(Field.DESCRIPTOR, Field)
+                this.register(CompositeType.DESCRIPTOR, CompositeType)
+                this.register(Choice.DESCRIPTOR, Choice)
+                this.register(RestrictedType.DESCRIPTOR, RestrictedType)
+                this.register(ReferencedObject.DESCRIPTOR, ReferencedObject)
+                this.register(TransformsSchema.DESCRIPTOR, TransformsSchema)
+                this.register(TransformTypes.DESCRIPTOR, TransformTypes)
```
I'm sure you don't need to prefix `register` with `this.`.

chrisr3 reviewed (commented):
I haven't tried checking this branch out, but I suppose it's OK. The DJVM's own serialization tests _should_ provide test coverage of `serialization-deterministic` and its new `LazyPool`.

rick-r3 committed:
PR review feedback (and removed TODO)

rick-r3 committed:
Make performance test look at wire transaction

rick-r3 committed:
Reduce iterations in performance test (but JIT seems to be over)

rick-r3 committed:
More PR feedback

rick-r3 committed:
More PR feedback

rick-r3 committed:
Fix detekt warnings

rick-r3 committed:
Fix lint warnings

rick-r3 commented:
@chrisr3 I'm getting an error from ProGuard.  It looks like the `position(int)` method is overloaded in J11 to return the narrowed `ByteBuffer` type, whereas not in J8 (it's on superclass `Buffer` and returns that).  Maybe this task is somehow not looking at J11 implementation.

```
[2023-07-16T08:37:22.494Z] > Task :serialization-deterministic:checkDeterminism FAILED
[2023-07-16T08:37:22.494Z] Warning: net.corda.serialization.internal.amqp.Envelope$FastPathConstructor: can't find referenced method 'java.nio.ByteBuffer position(int)' in library class java.nio.ByteBuffer
[2023-07-16T08:37:22.494Z] Warning: there were 1 unresolved references to library class members.
[2023-07-16T08:37:22.494Z]          You probably need to update the library versions.
[2023-07-16T08:37:22.494Z]          (http://proguard.sourceforge.net/manual/troubleshooting.html#unresolvedlibraryclassmember)
[2023-07-16T08:37:22.494Z] 
[2023-07-16T08:37:22.494Z] FAILURE: Build failed with an exception.
```

rick-r3 commented:
Ah, yes, I forget that maybe this is all JDK 8 only for DJVM.  That task depends on `deterministic_rt_jar` which looks to be JDK8 only.  

Maybe I need to set source compatibility for serialisation to 1.8.

rick-r3 committed:
Attempt to fix proguard complaint about missing JDK11 specific method

chrisr3 commented on serialization/build.gradle:
```diff
@@ -8,6 +8,7 @@ description 'Corda serialization'
 
 // required by DJVM and Avian JVM (for running inside the SGX enclave) which only supports Java 8.
 targetCompatibility = VERSION_1_8
+sourceCompatibility = VERSION_1_8
```
That looks like something that would already have been there?!

But if `serialization/build.gradle` needs that then I would expect `core/build.gradle` does too.

rick-r3 replied:
I'm speculating on the solution.  My local build isn't working and it seems the Jenkins one isn't either.  I think the problem is that the module is compiled against J11 libs which have different API to J8, which then linked with J8 causes issues because the generated code is not binary compatible with J8.  Not yet sure on solution.

chrisr3 replied:
Just checking - are the `JavaCompile` and `KotlinCompile` tasks for `core` and `serialization` generating Java 8 bytecode?

Or is Gradle complaining about not finding suitable artifacts for `core` and `serialization`?

rick-r3 replied:
Just pushed a change that appears to work locally, now that I don't have maven repo issues.

chrisr3 replied:
Do you mean this?
```
[2023-07-17T06:56:33.097Z] > Task :serialization-deterministic:checkDeterminism FAILED
[2023-07-17T06:56:33.097Z] Warning: net.corda.serialization.internal.amqp.Envelope$FastPathConstructor: can't find referenced method 'java.nio.ByteBuffer position(int)' in library class java.nio.ByteBuffer
```

chrisr3 replied:
FYI the `checkDeterminism` step validates that the reduced byte-code is consistent with the APIs inside `deterministic-rt.jar`, but the problem here seems to be that the return type of `ByteBuffer.position(int)` is `Buffer` on Java 8, whereas its `ByteBuffer` on Java 11.

Which is a relief - I was worried that I'd deleted `ByteBuffer.position(int)` from `deterministic-rt.jar`! :disappointed_relieved:

rick-r3 replied:
Yes.   I have coerced the type back up to the super class `Buffer` so it doesn't compile against the new J11 API.  Seems to work.  All passing now.

rick-r3 committed:
Attempt to fix proguard complaint about missing JDK11 specific method

(also checking in Constants.kt because it gets modified by IntelliJ every time I touch the code)

rick-r3 commented:
J11 vs J8 compatibility resolved

adelel1 commented on serialization/src/main/kotlin/net/corda/serialization/internal/amqp/Envelope.kt:
```diff
@@ -27,6 +37,7 @@ data class Envelope(val obj: Any?, val schema: Schema, val transformsSchema: Tra
         private const val TRANSFORMS_SCHEMA_IDX = 2
 
         fun get(data: Data): Envelope {
```
Is this function used anywhere?

rick-r3 replied:
That's a good question.  Let me have a look.

rick-r3 replied:
Only in tests.  Migrated 2 tests to use lazy initialisation for Envelope.

adelel1 commented on serialization/src/main/kotlin/net/corda/serialization/internal/amqp/Envelope.kt:
```diff
@@ -61,12 +72,59 @@ data class Envelope(val obj: Any?, val schema: Schema, val transformsSchema: Tra
+            val buffer = _buffer
+            val size = readEncodingAndReturnSize(buffer, false)
+            val data = Data.Factory.create()
+            data.decode(buffer)
```
The getEnvelope() in the original version raised an exception if there was an unexpected size of data, with corrupted blob. Is that needed here?

rick-r3 replied:
Yes.   I had it at one point but seem to have lost it.  Let me put it back.

adelel1 commented on serialization/src/main/kotlin/net/corda/serialization/internal/amqp/DeserializationInput.kt:
```diff
@@ -72,17 +74,32 @@ class DeserializationInput constructor(
+                decoderPool.reentrantRun {
+                    it.byteBuffer = dataBytes
+                    (it.readObject() as Envelope).apply {
+                        if (!lazy) this.resolvedSchema
```
Does the performance enhancement rely on the fact that the lambda is never actually called? 
I don't see where we parse it and store the results of that parse for all future envelopes. As its resolved individually for each Envelope.

rick-r3 replied:
Sort of.  The two properties on `Envelope`of `schema` and `transformsSchema` that are made lazy here are only requested if a serializer for the fingerprint of a class is not found in the `serializerFactory` `ConcurrentHashMap` of serialisers it has previously required for serialisation or deserialisation.  So essentially the `serializerFactory` has a memory, and over time will stop asking for `schema` or `transformSchema` as it will have seen all the classes.  That bit hasn't changed - this is just capitalising on it.

(BTW, the `lazy` flag here just forces that to be resolved for tests that expect the `Envelope` to be returned for some assertions, and we can't perform the lazy operations once we have handed the parser back to the pool of parsers - and the parser has expensive init so we don't want to create one for each deserialise).

rick-r3 committed:
More PR feedback
